File: utils/aoc_utils.py
import time
import tracemalloc
  
def aoc_run(expr):
    tracemalloc.start()
    
    start_time = time.time()
    
    # Run the function
    result = eval(expr)
    
    end_time = time.time()
    
    current, peak = tracemalloc.get_traced_memory()
    
    tracemalloc.stop()
    
    elapsed = end_time - start_time
    
    if elapsed < 0.001:
      elapsed = "< 0.001"
    else:
      elapsed = round(elapsed, 3)
      
    mem_used = peak / 1024
    
    if mem_used < 1:
      mem_used = "< 1"
    else:
      mem_used = round(mem_used)
    
    print("Elapsed:", elapsed, "seconds")
    print("Memory: ", mem_used, "KB")
    
    return elapsed, mem_used

def aoc_read(day):
  day = str(day)
  
  with open("input/day" + day + ".txt") as file:
    return file.readlines()
  
def aoc_source(day, part):
  day = str(day)
  part = str(part)
  
  with open("python/solve_day" + day + "_part" + part + ".py") as f:
    exec(f.read(), globals())

File: utils/aoc_utils.R
aoc_read <- function(day) {
  path <- paste0("input/day", day, ".txt")
  readLines(path)
}

aoc_source <- function(day, part) {
  path <- paste0("R/solve_day", day, "_part", part, ".R")
  source(path)
}

aoc_run <- function(expr) {
  expr <- substitute(expr)

  result <- suppressWarnings(
    bench::mark(eval(expr), iterations = 1)
  )

  time <- unclass(result[["total_time"]])

  if (time < 0.001) {
    time <- "< 0.001"
  } else {
    time <- round(time, 3)
  }

  mem_alloc <- unclass(result[["mem_alloc"]]) / 1024

  if (mem_alloc < 1) {
    mem_alloc <- "< 1"
  } else {
    mem_alloc <- round(mem_alloc)
  }

  cat("Elapsed:", time, "seconds\n")
  cat("Memory: ", mem_alloc, "KB\n")

  invisible(list(time, mem_alloc))
}
